api/buckets/service.go

```diff
@@ -616,36 +739,74 @@ func (s *Service) SetPath(ctx context.Context, req *pb.SetPathRequest) (*pb.SetP
 	}
 	dbToken, _ := thread.TokenFromContext(ctx)
 
-	buck := &tdb.Bucket{}
-	err := s.Buckets.Get(ctx, dbID, req.Key, buck, tdb.WithToken(dbToken))
+	destPath := cleanPath(req.Path)
+	bootCid, err := cid.Decode(req.Cid)
 	if err != nil {
+		return nil, fmt.Errorf("invalid remote cid: %s", err)
+	}
+
+	buck := &tdb.Bucket{}
+	if err = s.Buckets.GetSafe(ctx, dbID, req.Key, buck, tdb.WithToken(dbToken)); err != nil {
 		return nil, fmt.Errorf("get bucket: %s", err)
 	}
-	buckPath := path.New(buck.Path)
 
-	remoteCid, err := cid.Decode(req.Cid)
+	buck.UpdatedAt = time.Now().UnixNano()
+	buck.SetMetadataAtPath(destPath, tdb.Metadata{
+		UpdatedAt: buck.UpdatedAt,
+	})
+	buck.UnsetMetadataWithPrefix(destPath + "/")
+
+	txn, err := s.Buckets.WriteTxn(ctx, dbID, tdb.WithToken(dbToken))
 	if err != nil {
-		return nil, fmt.Errorf("invalid remote cid: %s", err)
+		return nil, err
+	}
+	if err = txn.Verify(ctx, buck); err != nil {
```

Comment:
Here's this new `Verify` method in use. The idea is that we should not change the underlying UnixFs bucket directory if the final bucket save is going to fail. Instead,
- Create a transaction
- Check if the change will success (if the token has permission)
- Perform the UnixFs changes
- Finally, save the bucket and end the transaction (commit the changes)

Still, this is not perfect... any non-access related errors during save will leave the relationship between the UnixFs directory and the bucket state broken. Though, this is not different than the current setup.